client/ayon_deadline/repository/custom/plugins/UnrealEngine5/JobPreLoad.py

```diff
@@ -0,0 +1,207 @@
+        deadlinePlugin.LogWarning(str(argError))
+        deadlinePlugin.LogWarning("Changelist will be latest in subsequent commands.")
+
+
+    #
+    # Sync project
+    #
+    if bSyncProject:
+
+        # Estimate how much work there is to do for a sync operation.
+        try:
+            deadlinePlugin.SetStatusMessage("Estimating work for Project sync (CL %d)" % perforceTools.changelist)
+            deadlinePlugin.LogInfo("Estimating work for Project sync (CL %d)" % perforceTools.changelist)
+            perforceTools.DetermineSyncWorkEstimate(bForceFullSync)
+        except UnrealSyncUtil.PerforceResponseError as argError:
+            deadlinePlugin.LogWarning(str(argError))
+            deadlinePlugin.LogWarning("No sync estimates will be available.")
+
+        # If there's no files to sync, let's skip running the sync. It takes a lot of time as it's a double-estimate.
+        if perforceTools.syncEstimates[0] == 0 and perforceTools.syncEstimates[1] == 0 and perforceTools.syncEstimates[2] == 0:
+            deadlinePlugin.LogInfo("Skipping sync command as estimated says there's no work to sync!")
+        else:
+            # Sync to the changelist already calculated.
+            try:
+                deadlinePlugin.SetStatusMessage("Syncing to CL %d" % perforceTools.changelist)
+                deadlinePlugin.LogInfo("Syncing to CL %d" % perforceTools.changelist)
+                deadlinePlugin.SetProgress(0)
+                deadlinePlugin.LogInfo("Estimated Files %s (added/updated/deleted)" % ("/".join(map(str, perforceTools.syncEstimates))))
+
+                logCallback = lambda tools: deadlinePlugin.SetProgress(perforceTools.GetSyncProgress() * 100)
+
+                # Perform the sync. This could take a while.
+                perforceTools.Sync(logCallback, bForceFullSync)
+
+                # The estimates are only estimates, so when the command is complete we'll ensure  it looks complete.
+                deadlinePlugin.SetStatusMessage("Synced Workspace to CL " + str(perforceTools.changelist))
+                deadlinePlugin.LogInfo("Synced Workspace to CL " + str(perforceTools.changelist))
+                deadlinePlugin.SetProgress(100)
+            except IOError as ioError:
+                deadlinePlugin.LogWarning(str(ioError))
+                deadlinePlugin.FailRender("Suspected Out of Disk Error while syncing: \"%s\"" % str(ioError))
+    else:
+        deadlinePlugin.LogInfo("Skipping Project Sync due to job settings.")
+
+
+
+    #
+    # Build project
+    #
+    if bBuildProject:
+        # BuildUtils requires engine root to determine a path to UnrealBuildTool
+        # Using Deadline system to determine the path to the executable
+        version = deadlinePlugin.GetPluginInfoEntry("EngineVersion")
+        deadlinePlugin.LogInfo('Version defined: %s' % version )
+        version_string = str(version).replace(".", "_")
+        executable_key = f"UnrealEditorExecutable_{version_string}"
```

Comment:
So with removing of hardcoded Unreal executable path by b57f535265ea91ea26c1f6681be7bc98cce825a1 this needs basically `UnrealEditorExecutable_5_4` to be preset in the param (possibly `UnrealEngine5.param`) file, but currently there's nothing. Shouldn't we at least provide entries there for the latest 3 versions (5.2, 5.3 and 5.4)? Without it, it crashes:

```
Error: Job preload script 'C:\ProgramData\Thinkbox\Deadline10\workers\PHOBOS\plugins\66b25541b95e063dbc2eee1c\JobPreLoad.py': Unreal Engine 5.4 entry not found in .param file (FranticX.Scripting.PythonNetException)
   at Deadline.Plugins.DeadlinePlugin.FailRender(String message)
```

Reply:
Or even better - set `UnrealExecutable` environment variable using AYON environments?

Reply:
I am setting `UnrealExecutable` in env vars, see in Description
![image](https://github.com/user-attachments/assets/9c077e37-8333-4a9b-82ed-fb06e056c1f1)

Reply:
oh right, I am doing this too, but still getting thit error 🤔
There is even this entry in Job log:
```
2024-08-07 10:11:43:  0: INFO: Setting Process Environment Variable UnrealExecutable to C:\Program Files\Epic Games\UE_5.4\Engine\Binaries\Win64\UnrealEditor-Cmd.exe
```

Reply:
so it should take both `UnrealExecutable` or `UnrealEditorExecutable_5_4` (notice that extra _Editor_). So the problem is, that these variables will not make it to the UnrealEngine5 plugin proces (`os.getenv()` in _JobPreLoad.py_ doesn't have them)